File: C/02_Types/01_Contiguous/simple.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100
#define NITEMS 100

int main(int argc, char** argv)
{

  int rank, recv_rank, nproc, left, right, iitems, sum;
  int values[NITEMS], values_recv[NITEMS];

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  for (iitems = 0; iitems < NITEMS; ++iitems){
    values[iitems] = rank;
  }

  if (rank == 0) printf("Conventional sendrecv\n");

  MPI_Sendrecv(values, NITEMS, MPI_INT, right, TAG, values_recv, NITEMS,
      MPI_INT, left, TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

  sum = 0;
  for (iitems = 0; iitems < NITEMS; ++iitems){
    sum += values_recv[iitems];
  }

  printf("Rank %3d got message from rank %3d with total of %5d\n", rank,
      left, sum);

  MPI_Finalize();

}

File: Fortran/02_Types/01_Contiguous/simple.f90
PROGRAM simple

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100
  INTEGER, PARAMETER :: nitems = 100

  INTEGER :: rank, recv_rank
  INTEGER :: nproc
  INTEGER :: left, right
  INTEGER :: ierr
  INTEGER :: contig_type
  INTEGER, DIMENSION(nitems) :: values, values_recv

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Set up periodic domain
  left = rank - 1
  IF (left < 0) left = nproc - 1
  right = rank + 1
  IF (right > nproc - 1) right = 0

  values = rank

  IF (rank == 0) PRINT *,'Conventional sendrecv'
  CALL MPI_Sendrecv(values, nitems, MPI_INTEGER, right, tag, values_recv, &
      nitems, MPI_INTEGER, left, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
  PRINT *,"Rank ", rank, " got message from rank ", left, " with total of ", &
      SUM(values_recv)

  CALL MPI_Finalize(ierr)

END PROGRAM simple
